include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContributionApproximator.h

```diff
@@ -0,0 +1,397 @@
+/// has a very large potential contribution as many samples are required to establish that allmost all of the box is covered.
+///
+///\tparam Rng The type of the Rng for sampling random points.
+struct HypervolumeContributionApproximator{
+	/// \brief Models a point and associated information for book-keeping purposes.
+	template<typename VectorType>
+	struct Point {
+		Point( VectorType const& point, VectorType const& reference ) 
+		: point( point )
+		, sample( point.size() )
+		, boundingBox( reference )
+		, boundingBoxVolume( 0. )
+		, approximatedContribution( 0. )
+		, contributionLowerBound( 0. )
+		, contributionUpperBound( 0. )
+		, noSamples( 0 )
+		, noSuccessfulSamples( 0 )
+		{}
+
+		VectorType point;
+		VectorType sample;
+		VectorType boundingBox;
+		std::vector< typename std::vector<Point>::const_iterator > influencingPoints;
+
+		double boundingBoxVolume;
+		double approximatedContribution;
+		double contributionLowerBound;
+		double contributionUpperBound;
+
+		std::size_t noSamples;
+		std::size_t noSuccessfulSamples;
+	};
+
+	double m_startDeltaMultiplier;
+	double m_multiplierDelta;
+	double m_minimumMultiplierDelta;
+	
+	double m_gamma;
+	double m_errorProbability; ///<The error probability.
+	double m_errorBound;  ///<The error bound
+
+	/// \brief C'tor
+	/// \param [in] delta the error probability of the least contributor
+	/// \param [in] eps the error bound of the least contributor
+	HypervolumeContributionApproximator()
+	: m_startDeltaMultiplier( 0.1 )
+	, m_multiplierDelta( 0.775 )
+	, m_minimumMultiplierDelta( 0.2 )
+	, m_gamma( 0.25 )
+	, m_errorProbability(1.E-2)
+	, m_errorBound(1.E-2)
+	{}
+		
+	double delta()const{
+		return m_errorProbability;
+	}
+	double& delta(){
+		return m_errorProbability;
+	}
+	
+	double epsilon()const{
+		return m_errorBound;
+	}
+	double& epsilon(){
+		return m_errorBound;
+	}
+	
+	template<typename Archive>
+	void serialize( Archive & archive, const unsigned int version ) {
+		archive & BOOST_SERIALIZATION_NVP(m_startDeltaMultiplier);
+		archive & BOOST_SERIALIZATION_NVP(m_multiplierDelta);
+		archive & BOOST_SERIALIZATION_NVP(m_minimumMultiplierDelta);
+		archive & BOOST_SERIALIZATION_NVP(m_gamma);
+		archive & BOOST_SERIALIZATION_NVP(m_errorProbability);
+		archive & BOOST_SERIALIZATION_NVP(m_errorBound);
+	}
+
+	/// \brief Determines the point contributing the least hypervolume to the overall set of points.
+	///
+	/// \param [in] s pareto front of points
+	/// \param [in] reference The reference point to consider for calculating individual points' contributions.
+	template<class Set,class VectorType>
+	std::vector<KeyValuePair<double,std::size_t> > smallest(Set const& points, std::size_t k, VectorType const& reference)const{
+		SHARK_CHECK(k == 1, "not implemented for k != 1");
+				
+		std::vector< Point<VectorType> > front;
+		for(auto it = points.begin(); it != points.end(); ++it ) {
+			front.emplace_back( *it, reference );
+		}
```

Comment:
Can't this loop be written without iterators? Not sure because Set is a template argument, but it seems to me that the loop is exactly equivalent to:

    for (auto const& point : points) front.emplace_back(point, reference);